src/charm.py

```diff
@@ -155,23 +159,166 @@ def _on_munge_available(self, event):
             self.unit.status = BlockedStatus("Error restarting munge")
             event.defer()
 
+    @classmethod
+    def _update_defaults(cls, **kwargs: Optional[str]):
+        """Update the environment settings in the /etc/defaults/slurmdbd file.
+
+        Updates the slurmdbd defaults environment file to add the specified
+        defaults. For example, to add the MY_VAR environment variable invoke with
+
+            self._update_defaults(my_var="foobar")
+
+        The key will be converted to all uppercase prior to being written to
+        the specified file. To remove a setting, specify the value as None, e.g.:
+
+            self._update_defaults(my_var=None)
+
+        Note: variable names will be converted to upper case when being written
+        to the file and for comparison of keys.
+
+        Args:
+            **kwargs:
+                The environment variables that should be set or unset. The key
+                will be upper-cased for the environment variable.
+        """
+        with open("/etc/default/slurmdbd", "w+") as f:
+            updated_contents = []
+
+            keys_processed = set()
+
+            for line in f.readlines():
+                # Lines beginning with a hash are a comment. Don't process these
+                # but do add the line to the output buffer to preserve it.
+                if line.startswith("#"):
+                    updated_contents.append(line)
+                    continue
+
+                # Attempt to get the environment variable and split it on the =.
+                # If the line doesn't have an equals, then don't process the line
+                # but do add the line to the output buffer to preserve it.
+                line_parts = line.split("=", 1)
+                if len(line_parts) < 2:
+                    updated_contents.append(line)
+                    continue
+
+                var_name = line_parts[0].lower()
+                if var_name not in kwargs:
+                    updated_contents.append(line)
+                    continue
+
+                keys_processed.add(var_name)
+                # If explicitly None, then remove from the output buffer as
+                # the demand is to unset the variable.
+                if kwargs[var_name] is None:
+                    continue
+
+                updated_contents.append(f"{var_name.upper()}={kwargs[var_name]}\n")
+
+            for key, value in kwargs.items():
+                # Skip any keys already processed.
+                if key in keys_processed:
+                    continue
+                updated_contents.append(f"{key.upper()}={value}\n")
+
+            f.seek(0)
+            f.truncate()
+            f.writelines(updated_contents)
```

Comment:
I also added this `_update_defaults(...)` methods to the `SlurmdbdManager` since that class is what manages the internal slurmdbd daemon on the machine. _I also reworked the manager class since I think I can improve on the original implementation._

Reply:
> I feel like this is a little complicated for what we need at the moment. Unless I am mistaken, all we really need to do is generate a /etc/default/slurmdbd file that has the MYSQL_UNIX_PORT environment variable set. We could just regenerate the /etc/default/slurmdbd as we need to to. Here's an implementation that I had:

Generally speaking, I agree that this is more complicated than we need. I was trying to account for operationally, an environment variable may need to be set in the defaults file and this was attempting to preserve that (this is from historical perspective in that operators *will* muck with some things). Additionally, not all sites will actually need or run a highly available database and if there are changes to the relation as a result, this would remove the setting when its not applicable anymore. This can, of course, be simplified by just re-writing that file without the environment variable present - but again, I was trying to account for the fact that there are certain files on the system that the operator will touch in times of duress and not wanting to break that.

I'm happy to change this to simply writing the file out with our intended contents if that's the consensus.

Reply:
I've refactored the code to be part of `SlurmdbdManager` and renamed it to `set_environment_var`. I have left the environment file manipulation for the time being.